### DISCUSSION TITLE
Is there an initializer for required members

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#6416] [Q&A] [answered]
When using initializers I mean init or required the object must be created beforehand and the values are set afterwards, right, so initializer are always set after construction

```cs
struct Data
{
  public int Value {get; init;}
}

var v = new Data()
{
  Value = 10;
}
```

some value are calculated depending on the incoming values though. But there is no way to call the constructor afterwards, right

```cs
struct Data
{
  public int Value {get; init;}
  public int Value2 {get; }

  public Data()
  {
    // Value is still 0
    Value2 = Value * 10;
  }
}

var v = new Data()
{
  Value = 10;
}
```

I'm asking this because of course I could make

` public int Value2 => Value * 10;`

which might be a performance sensitive thing, so the question is if some method could be emitted that is still in the context of a constructor, meaning I could still modify readonly members?

**Comment by HaloFour** [comment]
Initializer blocks are currently compiled into a call to the constructor followed by separate assignments to each property:

```cs
var v = new Data() {
    Value = 10;
};

// compiles into
var $temp = new Data();
$temp.Value = 10;
var v = $temp;
```

There are no methods automatically invoked at the end of an initializer in which you could inject this logic.  Currently the best you could probably do in this specific case would be to have the accessor for the property set the other property to the appropriate value:

```cs
struct Data {
    private int _value;
    public int Value {
        get => _value;
        init {
            _value = value;
            Value2 = value * 10;
        }
    }

    public int Value2 { get; private set; }
}
```

**Reply by FaustVX** [reply]
And even simpler with the `semi-auto prop` proposal #140
```C#
struct Data {
    public int Value {
        get;
        init {
            field = value;
            Value2 = value * 10;
        }
    }

    public int Value2 { get; private set; }
}
```

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-04-27.md#primary-constructor-bodies-and-validators
https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-05-04.md#final-initializers
https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-07-20.md#future-record-direction
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-01.md#c-11-initialization-triage

**Reply by msedi** [reply]
Thanks for the answer. At least its on the list ;-)
